src/mame/trs/agvision.cpp

```diff
@@ -0,0 +1,404 @@
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("CLEAR") PORT_CODE(KEYCODE_HOME) PORT_CHAR(12, UCHAR_SHIFT_2)
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("BREAK") PORT_CODE(KEYCODE_END) PORT_CODE(KEYCODE_ESC) PORT_CHAR(27)
+	PORT_BIT(0x78, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("SHIFT") PORT_CODE(KEYCODE_LSHIFT) PORT_CODE(KEYCODE_RSHIFT) PORT_CHAR(UCHAR_SHIFT_1)
+
+	PORT_START("hup")
+	PORT_BIT(0x01, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_NAME("HUP") PORT_CODE(KEYCODE_TILDE) PORT_CHANGED_MEMBER(DEVICE_SELF, agvision_state, agvision_state::hang_up, 0)
+INPUT_PORTS_END
+
+//-------------------------------------------------
+//  DEVICE_INPUT_DEFAULTS_START( ag_modem )
+//-------------------------------------------------
+
+static DEVICE_INPUT_DEFAULTS_START( ag_modem )
+	DEVICE_INPUT_DEFAULTS( "RS232_RXBAUD", 0xff, RS232_BAUD_300 )
+	DEVICE_INPUT_DEFAULTS( "RS232_TXBAUD", 0xff, RS232_BAUD_300 )
+	DEVICE_INPUT_DEFAULTS( "RS232_DATABITS", 0xff, RS232_DATABITS_8 )
+	DEVICE_INPUT_DEFAULTS( "RS232_PARITY", 0xff, RS232_PARITY_NONE )
+	DEVICE_INPUT_DEFAULTS( "RS232_STOPBITS", 0xff, RS232_STOPBITS_1 )
+DEVICE_INPUT_DEFAULTS_END
+
+//-------------------------------------------------
+//  machine_config
+//-------------------------------------------------
+
+void agvision_state::agvision(machine_config &config)
+{
+	this->set_clock(XTAL(14'318'181) / 16);
+
+	// basic machine hardware
+	MC6809E(config, m_maincpu, DERIVED_CLOCK(1, 1));
+	m_maincpu->set_addrmap(AS_PROGRAM, &agvision_state::agvision_mem);
+
+	PIA6821(config, m_pia_0, 0);
+	m_pia_0->readpa_handler().set(FUNC(agvision_state::pia0_pa_r));
+	m_pia_0->irqa_handler().set_inputline(m_maincpu, M6809_FIRQ_LINE);
+	m_pia_0->cb2_handler().set(FUNC(agvision_state::pia0_cb2_w));
+
+	// video hardware
+	SCREEN(config, m_screen, SCREEN_TYPE_RASTER).set_raw(3.579545_MHz_XTAL * 2, 456, 0, 320, 262, 0, 240);
+
+	MC6847_NTSC(config, m_vdg, XTAL(14'318'181) / 4); // VClk output from MC6883
+	m_vdg->set_screen(m_screen);
+	 m_vdg->hsync_wr_callback().set(m_pia_0, FUNC(pia6821_device::ca1_w));
+	m_vdg->input_callback().set(FUNC(agvision_state::sam_read));
+
+	// memory controller
+	SAM6883(config, m_sam, XTAL(14'318'181), m_maincpu);
+	m_sam->set_addrmap(2, &agvision_state::agvision_rom);			// ROM at $A000
+	 m_sam->set_addrmap(3, &agvision_state::agvision_static_ram);	// RAM at $C000
+	m_sam->set_addrmap(4, &agvision_state::agvision_io0);			//  IO at $FF00
+	m_sam->set_addrmap(5, &agvision_state::agvision_io1);			//  IO at $FF20
+	m_sam->set_addrmap(7, &agvision_state::agvision_boot);			//  IO at $FF60
+
+	RS232_PORT(config, m_rs232, default_rs232_devices, "null_modem");
+	m_rs232->set_option_device_input_defaults("null_modem", DEVICE_INPUT_DEFAULTS_NAME(ag_modem));
+
+	// internal ram
+	RAM(config, m_ram).set_default_size("16K").set_extra_options("39K,4K");;
```

Comment:
39K?  Really?  What arrangement of chips is that?

Reply:
The machine is advertised as coming in 4K and 16k models. But the software was written to look for RAM up to $9FFF. So I added that option. Admittedly it is a weird value. The only thing I can think of is a hand made development machine had that much RAM. On second thought, I'm changing it to 32K because the software never puts the machine into "All RAM" mode, so the extra bytes are always inaccessible.